compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java

```diff
@@ -322,89 +323,161 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
+                        selectType = types.getTypeIntersection(
+                                Types.IntersectionContext.typeTestIntersectionCalculationContext(),
+                                selectType, inferredSelectType, env);
+                    } else {
+                        BType checkedType = symTable.semanticError;
+                        if (selectType == symTable.semanticError) {
+                            checkedType = checkExpr(selectExp, env, data);
+                        }
+
+                        if (checkedType != symTable.semanticError && expTypes.size() == 1) {
+                            selectType = checkedType;
+                        }
+                    }
+
+                    if (queryExpr.isMap) { // A query-expr that constructs a mapping must start with the map keyword.
+                        resolvedType = symTable.mapType;
+                    } else {
+                        resolvedType = getNonContextualQueryType(selectType, collectionType);
+                    }
+                    break;
+            }
+            if (selectType != symTable.semanticError) {
+
+                if (resolvedType.tag == TypeTags.STREAM) {
+                    queryExpr.isStream = true;
                 }
-                if (queryExpr.isMap) { // A query-expr that constructs a mapping must start with the map keyword.
-                    resolvedType = symTable.mapType;
-                } else {
-                    resolvedType = getNonContextualQueryType(selectType, collectionType);
+                if (resolvedType.tag == TypeTags.TABLE) {
+                    queryExpr.isTable = true;
                 }
-                break;
-        }
-        if (selectType != symTable.semanticError) {
-            if (resolvedType.tag == TypeTags.STREAM) {
-                queryExpr.isStream = true;
+                possibleSelectTypes.add(selectType);
+                possibleResolvedTypes.add(resolvedType);
             }
-            if (resolvedType.tag == TypeTags.TABLE) {
-                queryExpr.isTable = true;
+        }
+        if (!possibleSelectTypes.isEmpty() && !possibleResolvedTypes.isEmpty()) {
+            selectTypes.addAll(possibleSelectTypes);
+            resolvedTypes.addAll(possibleResolvedTypes);
+        } else if (!errorTypes.isEmpty()) {
+            if (errorTypes.size() > 1) {
+                BType actualQueryType = silentTypeCheckExpr(queryExpr, symTable.noType, data);
+                if (actualQueryType != symTable.semanticError) {
+                    types.checkType(queryExpr, actualQueryType,
+                            BUnionType.create(null, new LinkedHashSet<>(expTypes)));
+                    errorTypes.forEach(expType -> {
+                        if (expType.tag == TypeTags.UNION) {
+                            checkExpr(selectExp, env, expType, data);
+                            selectExp.typeChecked = false;
```

Comment:
Can't we clone and do this instead? Are we sure there's nothing else that needs to be reset?

To avoid such issues and side-effects we usually clone and type check (when checking for compatibility). Check mapping constructor type checking against unions.

Reply:
Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/40835/commits/a587d82d3f1e1e27476f30f7d57d13928129df18).